doc/developer/arrangements.md

```diff
@@ -0,0 +1,98 @@
+We create more arrangements that just those to house materialized sources and views.
+Many internal operators require arrangements so that they can respond efficiently to changes.
+For example, the differential dataflow `join` operator requires that each of its two inputs be an arrangement.
+Additionally, the differential dataflow `reduce` operator requires that both its input and its output be an arrangement.
+These two operators are at the heart of the arrangement use in Materialize, because we construct relational operators out of them.
+
+Materialize has several relational operators that are "rendered" using differential `join` and `reduce` operators.
+These are the operators you would see in `explain plan for <query>`.
+There is a more complicated mapping from SQL to these operators.
+
+---
+
+Our `Join` operator joins together a sequence of input relations.
+It creates a sequence of differential `join` operators, each of which need to have both inputs be arrangements.
+This means that in addition to arranging each of the join inputs, it arranges each of the intermediate results.
+For example, in the figure below, each `A` is an arrangement.
+
+```
+    In2 --> A  In3 --> A  In4 --> A
+             \          \          \
+    In1 --> A-Join --> A Join --> A-Join --> Out
+```
+
+To join four inputs together, there are six arrangements.
+Some of these input arrangements could be shared, but most likely the intermediate arrangements cannot.
+
+---
+
+Our `Reduce` operator builds a non-trivial dataflow subgraph (which may be subject to change).
+Informally, for aggregation we build its own dataflow.
+That dataflow may begin with a `reduce` operator to implement the `distinct` keyword.
+Then there is either
+    * a single `reduce` operator (for sums, counts, any, all), or
+    * a sequence of 16 `reduce` operators (for min and max), or
+    * a single `reduce` operator (for `jsonbagg` and others).
+Finally, last `reduce` collects each of the aggregates into one record for each key.
+There are special cases when there are few (zero or one) reductions.
+
+---
+
+Our `TopK` operator builds a sequence of 16 `reduce` operators.
+Each `reduce` operator applies the `TopK` logic over progressively coarser groups
+The final group is just by the group key.
+
+---
+
+Our `Threshold` operator creates one `reduce` operator.
+
+---
+
+Our `ArrangeBy` operator creates one arrangement.
+
+---
+
+For each of these relational operators, the associated memory footprint should determined by the sizes of the arranged collections.
+
+## Caveats: Shared Arrangements
+
+Arrangements may be shared between operators in the same dataflow, or across dataflows for published arrangements.
+
+In particular, for `Join` it is often the case that the join will be "on" columns that form a primary key for the relation, and for which we have built an arrangement. In this case, we are able to re-use the existing arrangement, and incur no additional memory use for this arrangement.
```

Comment:
in the section describing arrangements it seems like the arrangement is keyed on the 'data' itself. so here when we talk about an arrangement for columns that constitute a primary key does that mean

- an arrangement where the 'data' consists of only primary key columns and no other columns

- something else? [iiuc arrangements have a key value structure and we default to having keys and values both be the Row itself and so its possible that we could have a key with just the subset of columns that make a primary key, but have the value be the whole Row - not sure if thats what you meant]

Reply:
Good point. I left out the key-value structure at first thinking that we could bypass it, but it is important for understanding shared arrangements.